src/Control/Carrier/NonDet/Church.hs

```diff
@@ -1,4 +1,14 @@
 {-# LANGUAGE DeriveTraversable, FlexibleInstances, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+-- | Provides 'NonDetC', a carrier for 'NonDet' effects providing choice and failure.
+--
+-- It can be invoked with custom functions for choice, success, and failure ('runNonDet'), or it can delegate
+-- said operations to an 'Control.Applicative.Alternative' instance ('runNonDetA') or an underlying monoidal
+-- result ('runNonDetM'). Under the hood, it uses a Church-encoded structure and a binary tree to prevent
+-- the problems associated with a naïve list-based implementation.
```

Comment:
Again, I don’t think we should be providing a thorough overview of the handler functions here.